### PR TITLE
Support choose local invoke in runtime instead of startup

### PR DESCRIPTION
**Pull request by MieAh**
## What is the purpose of the change
Finish issue#11013.To support choose local invoke in runtime instead of startup

## Brief changelog
In this change, I added ScopeClusterInvoker, and the mode similar to MockClusterInvoker put the judgment of local call in the runtime. However, when initializing the subscription, the local call will be judged in the original way.

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by MieAh** [comment]
@AlbumenJ  PTAL

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java [hunk 1]
```diff
@@ -420,8 +420,9 @@ private Map<String, String> appendConfig() {
 
     @SuppressWarnings({"unchecked"})
     private T createProxy(Map<String, String> referenceParameters) {
+        URL curUrl;
         if (shouldJvmRefer(referenceParameters)) {
-            createInvokerForLocal(referenceParameters);
+            curUrl = createInvokerForLocal(referenceParameters);
```
Comment (AlbumenJ) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java:425]: Remove create local invoker here

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java (hunk 1 above)
Comment (MieAh) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java]: done

**Comment by AlbumenJ** [comment]
Please fix conflicts

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    @Override
+    public Result invoke(Invocation invocation) throws RpcException {
+        Boolean peer = (Boolean) getUrl().getAttribute(PEER_KEY);
+        if (peer != null && peer) {
+            return invoker.invoke(invocation);
+        }
+        String scope = getUrl().getParameter(SCOPE_KEY);
+        if (shouldInvokeInjvm(getUrl().getParameter(LOCAL_PROTOCOL), scope)) {
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:125]: `peer`, `scope` 和 `local` 这几个 key 的判断是不是可以提前到 init 做

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            return injvmInvoker.invoke(invocation);
+        }
+        return invoker.invoke(invocation);
+    }
+
+    private boolean shouldInvokeInjvm(String isInjvm, String scope) {
+        if (Boolean.TRUE.toString().equals(isInjvm)) {
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:132]: equalsIgnoreCase

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        if (isExported.get() && injvmInvoker == null) {
+            if (injvmInvoker == null) {
+                synchronized (createLock) {
+                    if (injvmInvoker == null) {
+                        injvmInvoker = createInjvmInvoker();
+                    }
+                }
+            }
+        }
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:177]: 这里的时序问题可能导致 invoke 的时候 NPE

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    @Override
+    public void onExporterChangeUnExport(Exporter<?> exporter) {
+        if (getUrl().getServiceKey().equals(exporter.getInvoker().getUrl().getServiceKey())
+            && exporter.getInvoker().getUrl().getProtocol().equals(LOCAL_PROTOCOL)) {
+            isExported.compareAndSet(true, false);
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:184]: 需要销毁已有的 injvmInvoker

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/listener/InjvmExporterListener.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+public class InjvmExporterListener extends ExporterListenerAdapter {
+
+    private final Map<String, ExporterChangeListener> exporterChangeListeners = new ConcurrentHashMap<>();
+
+    private final Map<String, Exporter<?>> exporters = new ConcurrentHashMap<>();
+
+    @Override
+    public void exported(Exporter<?> exporter) throws RpcException {
+        String serviceKey = exporter.getInvoker().getUrl().getServiceKey();
+        exporters.putIfAbsent(exporter.getInvoker().getUrl().getServiceKey(), exporter);
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/listener/InjvmExporterListener.java:34]: 需要考虑同一个服务多订阅的情况

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java (hunk 2 above)
Comment (MieAh) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:125]: done

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java (hunk 3 above)
Comment (MieAh) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:132]: done

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java (hunk 5 above)
Comment (MieAh) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:184]: done

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/listener/InjvmExporterListener.java (hunk 6 above)
Comment (MieAh) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/listener/InjvmExporterListener.java:34]: 这里不应该是只考虑本地调用吗，也不需要经过网络传输，InjvmExporter 只会被当前的服务引用所订阅，理应不会出现多个订阅者的情况。而且每个服务只会暴露一个 InjvmExporter 实例，因此也不存在同一个服务多个 InjvmExporter 实例的情况

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java (hunk 4 above)
Comment (MieAh) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java]: done

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/listener/InjvmExporterListener.java (hunk 6 above)
Comment (MieAh) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/listener/InjvmExporterListener.java]: done

**Comment by AlbumenJ** [comment]
image
Please fix failed test cases

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+            return;
+        }
+        if (Boolean.TRUE.toString().equalsIgnoreCase(isInjvm) || LOCAL_KEY.equalsIgnoreCase(getUrl().getParameter(SCOPE_KEY))) {
+            injvmFlag = true;
+        } else if (isInjvm == null) {
+            injvmFlag = isNotRemoteOrGeneric();
+        }
+        if (injvmFlag) {
+            protocolSPI = getUrl().getApplicationModel().getExtensionLoader(Protocol.class).getAdaptiveExtension();
+            injvmExporterListener = (InjvmExporterListener) getUrl().getApplicationModel().getExtensionLoader(ExporterListener.class).getExtension(LOCAL_PROTOCOL);
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:95]: Convert ExporterListener into a scope bean.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Comment by AlbumenJ** [comment]
@CrazyHZM @chickenlj PTAL

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java (hunk 7 above)
Comment (MieAh) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:95]: done

**Review by CrazyHZM — APPROVED** [review]
LGTM

**Review by AlbumenJ — APPROVED** [review]
LGTM.

We can add some apis like `RpcContext.getServiceContext().setLocalInvoke(false)` to judge if use local invoker for each invocation.

**Comment by MieAh** [comment]
> LGTM.
> 
> We can add some apis like `RpcContext.getServiceContext().setLocalInvoke(false)` to judge if use local invoker for each invocation.

You mean that we can provide such an api to change the calling behavior, right?

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        return invoker.invoke(invocation);
+    }
+
+    private boolean isNotRemoteOrGeneric() {
+        return !SCOPE_REMOTE.equalsIgnoreCase(getUrl().getParameter(SCOPE_KEY)) &&
+            !getUrl().getParameter(GENERIC_KEY, false);
+    }
+
+    private boolean isInjvmExported() {
+        Boolean localInvoke = RpcContext.getServiceContext().getLocalInvoke();
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:155]: `RpcContext.getServiceContext().getLocalInvoke()` here should only effect **one** invocation.
If `null`, ignore.
If `false`, force remote.
If `true`, force local.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java (hunk 8 above)
Comment (MieAh) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/ScopeClusterInvoker.java:155]: done

**Comment by AlbumenJ** [comment]
Please update related docs in apache/dubbo-website

**Pull request merged**